File: study/lua/src/app/scene/view/homeland/HomelandPraySignNode.lua
-- 神树祈福福签
-- Author: Liangxu
-- Date: 2019-8-2
local ViewBase = require("app.ui.ViewBase")
local HomelandPraySignNode = class("HomelandPraySignNode", ViewBase)
local HomelandConst = require("app.const.HomelandConst")
local HomelandHelp = require("app.scene.view.homeland.HomelandHelp")

local EFFECT_COLOR = {
    [3] = "effect_shenshu_chouqian_languang",
    [4] = "effect_shenshu_chouqian_ziguang",
    [5] = "effect_shenshu_chouqian_huangguang",
}

function HomelandPraySignNode:ctor(pos, callback)
    self._pos = pos
    self._callback = callback
    self._canClick = true
	local resource = {
		file = Path.getCSB("HomelandPraySignNode", "homeland"),
		binding = {
			_imageBack = {
				events = {{event = "touch", method = "_onClick"}}
			},
		}
	}

	HomelandPraySignNode.super.ctor(self, resource)
end

function HomelandPraySignNode:onCreate()
	self._imageMid:setVisible(false)
end

function HomelandPraySignNode:updateUI(data)
    self._data = data
    if data then
        self._imageBack:setVisible(false)
        self._imageFront:setVisible(true)
        local info = data:getConfig()
        self._textColor:setString(info.color_text)
        self._textName:setString(info.name)
        local strTime = ""
        if info.type == HomelandConst.TREE_BUFF_TYPE_3 then
            strTime = Lang.get("homeland_buff_duration")
        end
        self._textTime:setString(strTime)

        local template = string.gsub(info.description, "#%w+#", "$c103_%1$")
        local value = HomelandHelp.getValueOfBuff(info.value, info.equation)
	    local times = HomelandHelp.getTimesOfBuff(info.times, info.type)
        local formatStr = Lang.getTxt(template, {value = value, times = times})
		local params = {defaultColor = Colors.NORMAL_BG_ONE, defaultSize = 20}
        local richText = ccui.RichText:createRichTextByFormatString(formatStr, params)
        richText:setAnchorPoint(cc.p(0.5, 1))
        richText:ignoreContentAdaptWithSize(false)
        richText:setContentSize(cc.size(160,0))
        richText:formatText()
        self._nodeDes:removeAllChildren()
        self._nodeDes:addChild(richText)
    else
        self._imageBack:setVisible(true)
        self._imageFront:setVisible(false)
        G_EffectGfxMgr:createPlayGfx(self._nodeEffectBack, "effect_shenshu_chouqian_xiaoxingxing", nil, nil, nil)
    end
end

--进场特效
function HomelandPraySignNode:playEnterEffect()
    G_EffectGfxMgr:createPlayGfx(self._nodeEffectFront, "effect_shenshu_chouqianfaguang", nil, nil, nil)
end

--抽时的特效
function HomelandPraySignNode:playDrawEffect()
    G_EffectGfxMgr:createPlayGfx(self._nodeEffectFront, "effect_shenshu_chouqian_shanbai", nil, nil, nil)
    self._nodeEffectMid:setVisible(true)
    self._imageMid:setVisible(true)
    local color = self._data:getConfig().color
    local midEffectName = EFFECT_COLOR[color]
    if midEffectName then
        G_EffectGfxMgr:createPlayGfx(self._nodeEffectMid, midEffectName, nil, nil, nil)
    end
end

function HomelandPraySignNode:stopDrawEffect()
    self._nodeEffectMid:setVisible(false)
    self._imageMid:setVisible(false)
end

function HomelandPraySignNode:_onClick()
    if self._canClick and self._callback then
        self._callback(self._pos)
    end
end

function HomelandPraySignNode:setClickEnable(enable)
    self._canClick = enable
end

return HomelandPraySignNode

File: study/ts/script/scene/view/homeland/HomelandPraySignNode.ts
import UIHelper from "../../../utils/UIHelper";
import { Path } from "../../../utils/Path";
import { HomelandConst } from "../../../const/HomelandConst";
import { Lang } from "../../../lang/Lang";
import { HomelandHelp } from "./HomelandHelp";
import { Colors, G_EffectGfxMgr } from "../../../init";
import { stringUtil } from "../../../utils/StringUtil";
import { RichTextExtend } from "../../../extends/RichTextExtend";

const { ccclass, property } = cc._decorator;

var EFFECT_COLOR = {
    [3]: 'effect_shenshu_chouqian_languang',
    [4]: 'effect_shenshu_chouqian_ziguang',
    [5]: 'effect_shenshu_chouqian_huangguang'
};

@ccclass
export default class HomelandPraySignNode extends cc.Component {

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBack: cc.Sprite = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeEffectBack: cc.Node = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageMid: cc.Sprite = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeEffectMid: cc.Node = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageFront: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textColor: cc.Label = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textName: cc.Label = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeDes: cc.Node = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textTime: cc.Label = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeEffectFront: cc.Node = null;
    _pos: any;
    _callback: any;
    _canClick: boolean;
    _data: any;


    ctor(pos, callback) {
        this._pos = pos;
        this._callback = callback;
        this._canClick = true;
        UIHelper.addEventListenerToNode(this.node, this._imageBack.node, 'HomelandPraySignNode', '_onClick');
    }
    onLoad() {
        this._imageMid.node.active = (false);
    }
    updateUI(data) {
        this._data = data;
        if (data) {
            this._imageBack.node.active = (false);
            this._imageFront.node.active = (true);
            var info = data.getConfig();
            this._textColor.string = (info.color_text);
            this._textName.string = (info.name);
            var strTime = '';
            if (info.type == HomelandConst.TREE_BUFF_TYPE_3) {
                strTime = Lang.get('homeland_buff_duration');
            }
            this._textTime.string = (strTime);
            var template = stringUtil.gsub(info.description, '#%w+#', '$c103_%1$');
            var value = HomelandHelp.getValueOfBuff(info.value, info.equation);
            var times = HomelandHelp.getTimesOfBuff(info.times, info.type);
            var formatStr = Lang.getTxt(template, {
                value: value,
                times: times
            });
            var params = {
                defaultColor: Colors.NORMAL_BG_ONE,
                defaultSize: 20
            };
            var richText = RichTextExtend.createRichTextByFormatString(formatStr, params);
            richText.node.setAnchorPoint(cc.v2(0.5, 1));
            richText.maxWidth = 160;
            this._nodeDes.removeAllChildren();
            this._nodeDes.addChild(richText.node);
        } else {
            this._imageBack.node.active = (true);
            this._imageFront.node.active = (false);
            G_EffectGfxMgr.createPlayGfx(this._nodeEffectBack, 'effect_shenshu_chouqian_xiaoxingxing', null, null, null);
        }
    }
    playEnterEffect() {
        G_EffectGfxMgr.createPlayGfx(this._nodeEffectFront, 'effect_shenshu_chouqianfaguang', null, null, null);
    }
    playDrawEffect() {
        G_EffectGfxMgr.createPlayGfx(this._nodeEffectFront, 'effect_shenshu_chouqian_shanbai', null, null, null);
        this._nodeEffectMid.active = (true);
        this._imageMid.node.active = (true);
        var color = this._data.getConfig().color;
        var midEffectName = EFFECT_COLOR[color];
        if (midEffectName) {
            G_EffectGfxMgr.createPlayGfx(this._nodeEffectMid, midEffectName, null, null, null);
        }
    }
    stopDrawEffect() {
        this._nodeEffectMid.active = (false);
        this._imageMid.node.active = (false);
    }
    _onClick() {
        if (this._canClick && this._callback) {
            this._callback(this._pos);
        }
    }
    setClickEnable(enable) {
        this._canClick = enable;
    }
}